cfme/tests/cloud_infra_common/test_provisioning.py

```diff
@@ -378,38 +390,57 @@ def test_cloud_provision_from_template_with_attached_disks(
         recursive_update(inst_args, {'environment': {'availability_zone': provisioning("av_set")}})
 
-    device_name = "/dev/sd{}"
+    device_name = "vd{}"
     device_mapping = []
 
-    with provider.mgmt.with_volumes(1, n=disks) as volumes:
-        for i, volume in enumerate(volumes):
-            device_mapping.append((volume, device_name.format(chr(ord("b") + i))))
-        # Set up automate
+    volumes = provider.mgmt.volume_configurations(1, n=disks)
+
+    @request.addfinalizer
+    def delete_volumes():
+        for volume in volumes:
+            provider.mgmt.delete_volume(volume)
 
-        method = modified_request_class.methods.instantiate(name="openstack_PreProvision")
+    # Set up automate
+    for i, volume in enumerate(volumes, 0):
+        if i == 0:
+            provider.mgmt.capi.volumes.set_bootable(volume, True)
+            device_mapping.append(
+                {'boot_index': 0,
+                'uuid': volume,
+                'device_name': device_name.format(chr(ord("a") + i))})
+        else:
+            device_mapping.append(
+                {'boot_index': -1,
+                'uuid': volume,
+                'device_name': device_name.format(chr(ord("a") + i))})
 
+    method = modified_request_class.methods.instantiate(name="openstack_PreProvision")
+    with update(method):
+        disk_mapping = []
+        for mapping in device_mapping:
+            disk_mapping.append(ONE_FIELD.format(**mapping))
+        method.script = VOLUME_METHOD.format(",\n".join(disk_mapping))
+
+    @request.addfinalizer
+    def _finish_method():
         with update(method):
-            disk_mapping = []
-            for mapping in device_mapping:
-                disk_mapping.append(ONE_FIELD.format(*mapping))
-            method.script = VOLUME_METHOD.format(", ".join(disk_mapping))
-
-        def _finish_method():
-            with update(method):
-                method.script = """prov = $evm.root["miq_provision"]"""
-        request.addfinalizer(_finish_method)
-
-        instance = appliance.collections.cloud_instances.create(vm_name,
-                                                                provider,
-                                                                form_values=inst_args)
-
-        for volume_id in volumes:
-            soft_assert(vm_name in provider.mgmt.volume_attachments(volume_id))
-        for volume, device in device_mapping:
-            soft_assert(provider.mgmt.volume_attachments(volume)[vm_name] == device)
-        instance.mgmt.delete()  # To make it possible to delete the volume
+            method.script = """prov = $evm.root["miq_provision"]"""
+
+    instance = appliance.collections.cloud_instances.create(vm_name,
+                                                            provider,
+                                                            form_values=inst_args)
+
+    @request.addfinalizer
+    def delete_vm_and_wait_for_gone():
+        instance.mgmt.delete()
```

Comment:
Question: Why not use `cleanup_on_provider`?

Reply:
Because delete will raise an exception when VM that is expected to be there will not be there, so we would see this as test ERROR or FAIL

Reply:
Based on the docblock of `cleanup_on_provider`, don't we want to avoid such errors in test teardown? @digitronik can you confirm this?

Reply:
Personally, I would avoid to direct method of `mgmt` as we have some Wrapper methods. It will not raise `NotFoundError` just it will log in logger.

Reply:
@digitronik If the test intends to  create a VM, and the VM disappears during the test or if it were never created, then there was either some bug in the tests code or there is a problem somewhere which should be fixed. Or someone went and deleted the VM. In any case like that I don't see a problem to make the test state ERROR as it failed to cleanup what it intended to create.

This approach will increase the visibility of problems. Or do you think it is good to just put the problems under the carpet (to the log) and put on a ping glasses and pretend everything is fine?

Reply:
@jaryn

@john-dupuy and @digitronik have valid points here about the test strategy and failure modes here.

I don't think there's any need for the negative tone taken in your response, or fallacy of them intending to hide test failures.  I assure you they both have the best intent for _our_ success.

You have expressed in your comments an implicit assertion that you're relying on in calling this delete method directly.  This implicit assertion/assumption has additionally been placed in the shutdown phase of the test function, without any inline comments to the additional assertion that's in your test.

An assertion that the VM still exists at the end of the test function should be included in the call phase of the test function, not shutdown.

It should also be explicit in its mode of assertion - do you want to 1) assert that the VM is still visible in cloudforms, in a non-archived, non-orphaned state, 2) just still visible in cloudforms, 3) visible on the provider, 4) in a specific power state, etc.

Reply:
I am sorry that people might take this personally but I still don't understand why is `cleanup_on_provider` which is first checking for existence is any better than just `instance.mgmt.delete()`.

Reply:
OK. I changed it to cleanup_on_provider.